cmd/lakectl/cmd/fs_cp.go

```diff
@@ -0,0 +1,145 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"github.com/spf13/cobra"
+	"github.com/treeverse/lakefs/pkg/api/apigen"
+	"github.com/treeverse/lakefs/pkg/api/apiutil"
+	"github.com/treeverse/lakefs/pkg/uri"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"sync"
+)
+
+var fsCpCmd = &cobra.Command{
+	Use:   "cp <source URI> <destination URI>",
+	Short: "Copy object",
+	Args:  cobra.ExactArgs(2),
+	Run: func(cmd *cobra.Command, args []string) {
+		recursive := Must(cmd.Flags().GetBool(recursiveFlagName))
+		concurrency := Must(cmd.Flags().GetInt("concurrency"))
+		sourceURI := MustParsePathURI("source", args[0])
+		destinationURI := MustParsePathURI("destination", args[1])
+
+		if sourceURI.Repository != destinationURI.Repository {
+			Die("Can only copy files in the same repo", 1)
+		}
+
+		client := getClient()
+
+		ctx := cmd.Context()
+
+		if !recursive {
+			err := copyObject(ctx, client, sourceURI, destinationURI)
+			if err != nil {
+				DieErr(err)
+			}
+			return
+		}
+
+		success := true
+		var errorsWg sync.WaitGroup
+		errors := make(chan error)
+		errorsWg.Add(1)
+		go func() {
+			defer errorsWg.Done()
+			for err := range errors {
+				_, _ = fmt.Fprintln(os.Stderr, err)
+				success = false
+			}
+		}()
```

Comment:
In other places (for instance, fs_upload.go) we use a control flow which is inverted compared to this one:

- Spawn a goroutine, compute work items there, load onto the workers channel.
- In the main goroutine, read all errors from the errors channel.

This has the advantage of avoiding a waitgroup (the main goroutine can just count errors or nil returns; or if using errgroup it can just ask for the errors directly)